video.1
.TH "VIDEO" "1" "24 August 2008" "" ""
.SH "NAME"
video \- MythNetTV video processing module
.SH "DESCRIPTION"
 \fBmythnettv\fP is a RSS video aggregtor for MythTV.  It take 
 downloads video feeds and places them in the recording directory
 in MythTV, then makes them accessible with the rest of your 
 recordings.

 The MythNetTV video subsystem may be queried directly. This
 can be useful if you want to perform simple operations such
 as asking if a given video file needs to be transcoded before
 being imported into MythTV, or actually performing the
 transcode.

.SH "USAGE"

 length <path>       : return the length of a video in seconds
 info <path>         : return all available information for a
                       given video
 needstranscode <path>
                     : determine if a given file needs to be
                       transcoded before being imported into
                       MythTV
 transcode <path> <output path>
                     : transcode the file into a format
                       suitable for MythTV

.SH "CONTACT DETAILS"
Information about \fBmythnettv\fP is posted to \fBhttp://www.stillhq.com/mythtv/mythnettv/\fP regularly. There is a users mailing list at \fBhttp://lists.stillhq.com/listinfo.cgi/mythnettv-stillhq.com\fP

.SH "AUTHOR"
\fBmythnettv\fP is by Michael Still <mikal@stillhq.com>



ChangeLog
  - r292: Find orphan recording files, and a command to delete only the oldest recordings
  - r290: Fix video unit test (exception changed)
  - r289: Changed permissions of imported files to be RW for everyone. This was prompted by moving to MythBuntu and using a different user for MythTV than MythNetTV. Perhaps this should become a config option?
  - r287: Fix plugin unit tests
  - r284: Add COPYING.BSD file, which covers gflags only
  - r282: Tweaks to transcode verbose flag and fix broken exception inheritance
  - r280: Catch impossibly small files being created by transcoding, and fix some output capture bugs with transcodes
  - r278, r279: Better handling of filenames with spaces in them
  - r278: WMV3 needs transcoding
  - r279: Handle downloads which create a directory with just one file in it
  - r276: Cleaned up the draft download plugin interface, some unit tests for bittorrent download
  - r274: Enable autoexpire on recordings
  - r272: Added the unfail command
  - r270: Fix single quote escaping in strings, and add unit tests for quote escaping
  - r268: Release script now checks for updated version number
  - r266: Fix release number
  - r264: Add missing license files
  - r262: Bug fixes for refactored bittorrent plugin
  - r260: Refactor bittorrent support into its own plugin
  - r258: Enforce sensible modes on recording files

------ Unsorted ChangeLog entries above this line ------

Release 6 -> Release 7:

  Better testing:
    - r232: Simple script to create the unit testing database

  New features:
    - r210: Allow users to override the default bittorrent upload rate
    - r212: Provide more sync information when running in verbose mode
    - r216: Add the markunread command
    - r218: Make only one download attempt for a given show per 24 hour period
    - r220: Change download interval to one hour, update last attempt time more frequently
    - r222: Add --oldestfirst and --newestfirst flags, which alter NextDownloads behaviour
    - r224: Add flag which forces downloads to run, even if they've failed recently
    - r226: Don't mark a bittorrent download as complete just because the torrent file has been completely downloaded
    - r234: Use reasonable defaults for mysql configuration if no config file can be found
    - r236: Use reasonable defaults for mysql configuration if no config file can be found
    - r238: Use reasonable defaults for mysql configuration if no config file can be found

  Bug fixes:
    - r206: Make the output of "video.py length" more friendly
    - r208: Release man pages as well
    - r214: Fix a bug introduced in r210 which broke bittorrent downloads
    - r228: Fix a typo in the recordings_tool help text
    - r230: Fix a typo in the recordings_tool help text
    - r240: Reduce size of proxy name in the usage table to deal with key length limitations in MySQL
    - r242: The schema for the channel table changed in MythTV 0.21
    - r244: Fix typo
    - r246: Suppress repeated warnings of DB default use
    - r248: Bug fixes for linux.conf.au 2009 talk
    - r250: Include test runner script in release, prepare ChangeLog for release

Release 5 -> Release 6:
  Better testing:
  - r114: Unit tests for video.py
  - r148: Updated unit test now that we create more tables
  - r151: Start work on syndication unit tests
  - r153: Unit test for bad syndication dates, as well as more flexiblity in
    db connections needed for this unit test
  - r157: Nicer unit test failure output, and refactor GetVideoDir() into the
    utility module
  - r181: Slightly improve syndication unit tests
  - r183: Add flag parsing to unit tests, and fix some more flag name changes

  Better documentation:
  - r117: Added a man page for video.py -- I'm not 100% happy with its name
    though

  Bug fixes:
  - r114: Fixed a bug where the new filename for video transcode could be the
    same as the input filename, resulting in video corruption. This was found
    with one of the new video.py unit tests
  - r116: The logic for the --prompt flag was the wrong way around. Fixed.
  - r119: Nicer download status messages
  - r121: Handle 404s in feed updates better
  - r129: Slight tweak to SVN submit script
  - r131: More accurate tracking of proxy usage (update during download,
    instead of just at the end)
  - r137: Proxy budget being exceeded doesn't count as a failed download
    attempt
  - r143: Subscribe now renables inactive subscriptions
  - r146: Add support to decimals to utility byte printer, fix a bug in the
    check for video directories
  - r155: Have users send problems to the mailing list, instead of me
    personally
  - r161: Don't throw exceptions for the videodir command line
  - r167, 169: Display friendly sizes in records_tool output
  - r171: Move verbose update arg into a flag
  - r173: Add "-vo null" to mplayer invocation per Ryan Lutz. This improves
    support on machines without X, and speeds up the identify operation
  - r175: Import patch from Thomas Mashos which fixes subscription removal,
    started work on syndication unit test improvements
  - r177: Fix character escaping bug in show subtitles during import
  - r179: Renamed --datadirdefault to --datadir. If set this will now change
    your data directory, regardless of if there was a previous value set.
  - r190: Recording_tool now prompts for deletes
  - r192: Improved explainvideodir output
  - r194: Don't crash in explainvideodir if there is no video directory
  - r197: Order results by subtitle in recordings_tool output

  New features:
  - r115: Upgraded schema to version 15 to support http_proxies for
    subscriptions. Added http_proxy command line, which allows you to use HTTP
    proxies for specified URLs. Moved HTTP accesses to use the proxy code.
  - r127: Bump schema to v17, and add proxy use tracking including the
    "proxyusage" command
  - r133: Allow daily budgets for proxy usage
  - r115: Provide a user agent for HTTP requests, instead of just 

video.py
#!/usr/bin/python

# Copyright (C) Michael Still (mikal@stillhq.com) 2006, 2007, 2008, 2009
# Released under the terms of the GNU GPL v2
import commands
import datetime
import os
import re
import sys

from stat import *

import gflags
import utility

# Define command line flags
gflags.DEFINE_boolean('verbose_transcode', False,
                      'Output verbose debugging information')
FLAGS = gflags.FLAGS


# Exceptions returned by this module
class TranscodeException(utility.LoggingException):
  """ Transcoding problems """

class LengthException(Exception):
  """ Exception while determining the length of a video """

class ParseException(Exception):
  """ Exception while trying to parse video characteristics """


class MythNetTvVideo:
  """MythNetTvVideo -- video handling methods

  This uses mplayer to determine the length of the video. Specifically, this
  command line does the trick:

  mplayer -frames 0 -identify <filename> 2>&1 | grep ID_LENGTH
  """

  def __init__(self, db, filename):
    """__init__ -- prime the pump to answer questions about the video"""

    self.db = db
    self.filename = filename
    self.values = {}

    if not os.path.exists(self.filename):
      raise ParseException('Video file missing')

    out = commands.getoutput('mplayer -vo null -frames 0 -identify "%s" '
                             '2>&1 | grep "="' \
                             % self.filename)
    for line in out.split('\n'):
      try:
        (key, value) = line.split('=')
        self.values[key] = value
      except:
        pass

    if len(self.values) == 0:
      raise ParseException('Could not parse video characteristics')

  def Length(self):
    """Length -- return the length of the video in seconds"""
    if 'ID_LENGTH' in self.values:
      return float(self.values['ID_LENGTH'])
    
    raise LengthException('Could not determine length of %s. '
                          'Attributes found = %s'
                          %(self.filename, self.values))

  def NeedsTranscode(self, out=sys.stdout):
    """NeedsTranscode -- decide if a video needs transcoding before import"""

    # Doesn't need transcoding
    return_false = ['mp4v', '0x10000002', 'divx', 'DIVX', 'XVID',
                    'DX50']

    # Does need transcoding
    # Note that some avc1 videos work, and some don't -- so all get transcoded
    return_true = ['FLV1']

    if self.values['ID_VIDEO_FORMAT'] in return_false:
      out.write('Files in format %s don\'t need transcoding\n'
                % self.values['ID_VIDEO_FORMAT'])
      return False

    if self.values['ID_VIDEO_FORMAT'] in return_true:
      out.write('Files in format %s do need transcoding\n'
                % self.values['ID_VIDEO_FORMAT'])
      return True

    else:
      out.write("""
****************************************************************
I don't know if we need to transcode videos in %s format
I'm going to give it a go without, and it if doesn't work
please report it to mikal@stillhq.com
****************************************************************
"""
                % self.values['ID_VIDEO_FORMAT'])
      return False

  def NewFilename(self, datadir, extn, out=sys.stdout):
    """NewFilename -- determine what filename to use after transcoding"""

    # We only return the filename portion, not the path
    re_filename = re.compile('^(.+)/(.+?)$')
    m = re_filename.match(self.filename)
    if m:
      file = m.group(2)
    else:
      file = self.filename

    # Try changing the extension to extn
    re_extension = re.compile('(.*)\.(.*?)')
    m = re_extension.match(file)
    if m:
      count = 1
      proposed_filename = '%s.%s' % (m.group(1), extn)
      while os.path.exists('%s/%s' %(datadir, proposed_filename)):
        out.write('Rejected %s as it already exists\n' % proposed_filename)
        count += 1
        proposed_filename = '%s_%d.%s' %(m.group(1), count, extn)

      return proposed_filename

    return 'new-%s' % file

  def Transcode(self, datadir, out=sys.stdout):
    """Transcode -- transcode the video to a better format. Returns the new
    filename.
    """

    # If the file is small, go for a format which will hopefully look nicer
    out.write('Transcoding\n')
    format = '-ovc lavc -oac lavc -lavcopts abitrate=128 -ffourcc DX50'
    newfilename = self.NewFilename(datadir, 'avi', out=out)
    start_size = os.stat(self.filename)[ST_SIZE]

    command = 'mencoder "%s" %s -o "%s/%s"' %(self.filename, format,
                                              datadir, newfilename)
    (status, output) = commands.getstatusoutput(command)

    if status != 0:
      raise TranscodeException(self.db,
                               'Transcode failed: %s\n%s\nCommand: %s'
                               %(status, output, command))

    # mencoder sometimes just returns junk data instead of doing the transcoding
    # -- the one example of this I have seen is when the file in question was
    # encumbered with DRM
    if os.stat('%s/%s' %(datadir, newfilename)).st_size < 10000:
      raise TranscodeException(self.db,
                               'Transcode failed: an impossibly small file '
                               'was returned')

    if FLAGS.verbose_transcode:
      out.write('----------\nmencoder output:\n%s\n----------\n\n' % output)

    # Log the file growth
    if self.db:
      self.db.Log('Transcoding changed size of file from %d to %d' \
                   %(start_size,
                   os.stat('%s/%s' %(datadir, newfilename))[ST_SIZE]))

    return newfilename


def Usage():
  print """Unknown command line. Try one of:'

The MythNetTV video subsystem may be queried directly. This
can be useful if you want to perform simple operations such
as asking if a given video file needs to be transcoded before
being imported into MythTV, or actually performing the
transcode.

length <path>       : return the length of a video in seconds
info <path>         : return all available information for a
                      given video
needstranscode <path>
                    : determine if a given file needs to be
                      transcoded before being imported into
                      MythTV
transcode <path> <output path>
                    : transcode the file into a format
                      suitable for MythTV
"""

  print '\n\nAdditionally, you can use these global flags:%s' % FLAGS
  sys.exit(1)


if __name__ == '__main__':
  # Parse flags
  try:
    argv = FLAGS(sys.argv)
  except gflags.FlagsError, e:
    out.write('%s\n' % e)
    Usage(out)

  # Present a simple user interface to query the video subsystem
  filename = None
  try:
    filename = argv[2]
    if not os.path.exists(filename):
      print '%s: file not found' % filename
      Usage()
  except:
    print 'Could not find a file argument'
    Usage()

  # Construct a video object
  vid = None
  try:
    vid = MythNetTvVideo(None, filename)
  except Exception, e:
    print 'Video processing error: %s' % e
    sys.exit(1)

  if argv[1] == 'length':
    try:
      print 'Length of %s: %s' %(sys.argv[2],
                                 utility.DisplayFriendlyTime(vid.Length()))
    except Exception, e:
      print 'Length error: %s' % e
      sys.exit(1)

  elif argv[1] == 'info':
    for key in vid.values:
      print '%s: %s' %(key, vid.values[key])

  elif argv[1] == 'needstranscode':
    print 'Needs transcoding: %s' % vid.NeedsTranscode()

  elif argv[1] == 'transcode':
    if vid.NeedsTranscode():
      print ('Created output file: %s/%s'
             %(argv[3], vid.Transcode(argv[3])))

  else:
    print 'Unknown command'
    Usage()
